Sources/ComposableArchitecture/Reducer.swift

```diff
@@ -290,39 +290,162 @@ public struct Reducer<State, Action, Environment> {
+  ///             // But the child cannot process this action if its state is `nil` in the parent
+  ///           ...
+  ///           }
+  ///         }
+  ///
+  ///     It is perfectly reasonable to ignore the result of an effect when child state is `nil`,
+  ///     for example one-off effects that you don't want to cancel. However, many long-living
+  ///     effects _should_ be explicitly canceled when tearing down a child domain:
+  ///
+  ///         let childReducer = Reducer<
+  ///           ChildState, ChildAction, ChildEnvironment
+  ///         > { state, action environment in
+  ///           struct MotionId: Hashable {}
+  ///
+  ///           switch action {
+  ///           case .onAppear:
+  ///             // Mark long-living effects that shouldn't outlive their domain cancellable
+  ///             return environment.motionClient
+  ///               .start()
+  ///               .map(ChildAction.motion)
+  ///               .cancellable(id: MotionId())
+  ///
+  ///           case .onDisappear:
+  ///             // And explicitly cancel them when the domain is torn down
+  ///             return .cancel(id: MotionId())
+  ///           ...
+  ///           }
+  ///         }
+  ///
+  ///   * A view store sends a child action when child state is `nil`:
+  ///
+  ///         WithViewStore(self.parentStore) { parentViewStore in
+  ///           // If child state is `nil`, it cannot process this action.
+  ///           Button("Child Action") { parentViewStore.send(.child(.action)) }
+  ///           ...
+  ///         }
+  ///
+  ///     Use `Store.scope` with`IfLetStore` or `Store.ifLet` to ensure that views can only send
+  ///     child actions when the child domain is non-`nil`.
+  ///
+  ///         IfLetStore(
+  ///           self.parentStore.scope(state: { $0.child }, action: { .child($0) }
+  ///         ) { childStore in
+  ///           // This destination only appears when child state is non-`nil`
+  ///           WithViewStore(childStore) { childViewStore in
+  ///             // So this action can only be sent when child state is non-`nil`
+  ///             Button("Child Action") { childViewStore.send(.action) }
+  ///           }
+  ///           ...
+  ///         }
   ///
   /// - See also: `IfLetStore`, a SwiftUI helper for transforming a store on optional state into a
   ///   store on non-optional state.
   /// - See also: `Store.ifLet`, a UIKit helper for doing imperative work with a store on optional
   ///   state.
-  public func optional(_ file: StaticString = #file, _ line: UInt = #line) -> Reducer<
+  ///
+  /// - Parameter breakpointOnNil: Raises `SIGTRAP` signal when an action is sent to the reducer
+  ///   but state is `nil`. This is generally considered a logic error, as a child reducer cannot
+  ///   process a child action for unavailable child state.
+  /// - Returns: A reducer that works on optional state.
+  public func optional(
+    breakpointOnNil: Bool = true,
```

Comment:
Thank you for making this change - I definitely understand the reasons for the previous strict behaviour, but I think this is a nice balance. Default to strict, with the option to ignore if needed.

I've run into a particular case recently with a NavigationLink's binding that sends a "dismiss" action after the child state is nullified, and the only way I've found to work around it is to put in a delay to let the navigation complete before nulling out the child state. Ignoring the error would be preferable to the delay. Though I acknowledge that there may be something incorrect about how I'm using the navigation link that is causing the problem.

Reply:
As always, thanks for the great work 👏 .

Would you consider a setup where we take a `canIgnoreAction: (Action) -> Bool = { _ in false }` parameter? Id love to retain the strict behavior for most actions and explicitly list those that can come in post mortem.